File: verifier-problems/src/main/java/com/jetbrains/pluginverifier/problems/Problem.java
package com.jetbrains.pluginverifier.problems;

import com.jetbrains.pluginverifier.persistence.Jsonable;
import org.jetbrains.annotations.NotNull;

import javax.xml.bind.annotation.XmlTransient;

/**
 * Problem container class.
 */
public abstract class Problem implements Jsonable<Problem> {

  @XmlTransient
  @NotNull
  //TODO: write a renderer for TC and others instead of prefix
  @Deprecated
  public abstract String getDescriptionPrefix();


  @XmlTransient
  @NotNull
  public abstract String getDescription();

  @Override
  public final boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || o.getClass() != getClass()) return false;
    return getDescription().equals(((Problem) o).getDescription());
  }

  @Override
  public final int hashCode() {
    return getDescription().hashCode();
  }

  @Override
  public final String toString() {
    return getDescription();
  }
}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/results/problems/Problem.kt
package com.jetbrains.pluginverifier.results.problems

/**
 * @author Sergey Patrikeev
 */
abstract class Problem {

  abstract val shortDescription: String

  abstract val fullDescription: String

  protected open val equalityReference: String
    get() = fullDescription

  final override fun toString(): String = fullDescription

  final override fun equals(other: Any?): Boolean = other is Problem && equalityReference == other.equalityReference

  final override fun hashCode(): Int = equalityReference.hashCode()

}